plat-sam: add sama5d2 clock tree support

clementleger opened the pull request:
This series add complete support for SAMA5D2 clock tree.
Drivers have been partially ported from Linux and have been relicensed with dual GPL2 and BSD-3-Clause.
This will need to be acked by @Noglitch and @bbrezillon.

jenswi-linaro commented on lib/libutils/ext/include/util.h:
```diff
@@ -75,6 +75,8 @@
 /* Round down the even multiple of size, size has to be a multiple of 2 */
 #define ROUNDDOWN(v, size) ((v) & ~((__typeof__(v))(size) - 1))
 
+#define DIV_ROUND_UP(x, y) (((x) + (y) - 1) / (y))
```
What's wrong with `ROUNDUP_DIV()`?

clementleger replied:
In the comments, in is indicated that ROUNDUP_DIV y parameter must be a value of 2. In my case it isn't since it's a rate to be set on the clock. If used with numbers which are non multiple of 2, it yields weird results which are non usable.

etienne-lms replied:
we should add a assertion on `y` value for `ROUNDUP_DIV()`.

jenswi-linaro replied:
OK, please give a proper description of the macro so one can understand when to use it and when to use `ROUNDUP_DIV()` instead.

jenswi-linaro commented:
I know we have two platforms with a drivers directory, but I'd prefer to add new drivers under `core/drivers/`.
In this case `core/drivers/sam` should make sense.

noglitch commented:
Acked-by: Nicolas Ferre <nicolas.ferre@microchip.com>

Thanks for this work. Best regards,
  Nicolas

clementleger commented:
> I know we have two platforms with a drivers directory, but I'd prefer to add new drivers under `core/drivers/`. In this case `core/drivers/sam` should make sense.

I'm fine with that !

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_peripheral.c:
```diff
@@ -0,0 +1,189 @@
+
+	if (periph->id < PERIPHERAL_ID_MIN || !periph->range.max) {
+		if (parent_rate == rate)
+			return 0;
```
TEE_SUCCESS

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_peripheral.c:
```diff
@@ -0,0 +1,189 @@
+		if (parent_rate == rate)
+			return 0;
+		else
+			return -1;
```
TEE_ERROR_GENERIC

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_pll.c:
```diff
@@ -0,0 +1,303 @@
+
+#include "at91_clk.h"
+
+#define PLL_STATUS_MASK(id)	(1 << (1 + (id)))
```
`BIT((id) + 1)`

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_pll.c:
```diff
@@ -0,0 +1,303 @@
+{
+	unsigned int status = io_read32(base + AT91_PMC_SR);
+
+	return status & PLL_STATUS_MASK(id) ? 1 : 0;
```
`return io_read32(base + AT91_PMC_SR) & PLL_STATUS_MASK(id);` does the job.

clementleger replied:
I guess this was a precaution for the bool cast also.

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_pll.c:
```diff
@@ -0,0 +1,303 @@
+
+	status = io_read32(pll->base + AT91_PMC_SR);
+	if ((status & mask) &&
+	    (div == pll->div && mul == pll->mul))
```
fits in a single line, maybe
```c
	if (status & mask && div == pll->div && mul == pll->mul)
		return TEE_SUCCESS;
```

clementleger replied:
Honestly, I'm not sure this is more readable.

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_plldiv.c:
```diff
@@ -0,0 +1,65 @@
+	struct pmc_data *pmc = clk->priv;
+
+	if ((parent_rate != rate) && (parent_rate / 2 != rate))
+		return -1;
```
`TEE_ERROR_GENERIC`

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_plldiv.c:
```diff
@@ -0,0 +1,65 @@
+	io_clrsetbits32(pmc->base + AT91_PMC_MCKR, AT91_PMC_PLLADIV2,
+			parent_rate != rate ? AT91_PMC_PLLADIV2 : 0);
+
+	return 0;
```
`TEE_SUCCESS`

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_sckc.c:
```diff
@@ -0,0 +1,54 @@
+#include <initcall.h>
+#include <io.h>
+#include <libfdt.h>
+#include <malloc.h>
```
reorder in alphabetical order

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_usb.c:
```diff
@@ -0,0 +1,122 @@
+#include "at91_clk.h"
+
+#define SAM9X5_USB_DIV_SHIFT	8
+#define	SAM9X5_USB_DIV_COUNT	BIT32(4)
```
fix indentation

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_usb.c:
```diff
@@ -0,0 +1,122 @@
+#define	SAM9X5_USB_DIV_COUNT	BIT32(4)
+#define SAM9X5_USB_MAX_DIV	(SAM9X5_USB_DIV_COUNT - 1)
+
+#define SAM9X5_USBS_MASK	GENMASK_32(0, 0)
```
`BIT(0)` ?

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_system.c:
```diff
@@ -0,0 +1,95 @@
+	uint8_t id;
+};
+
+static inline int is_pck(int id)
```
no `inline` functions in C source files.

should return a boolean

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_system.c:
```diff
@@ -0,0 +1,95 @@
+{
+	uint32_t status = io_read32(base + AT91_PMC_SR);
+
+	return !!(status & (1 << id));
```
`	return status & BIT(id);` is fine

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_pmc.c:
```diff
@@ -0,0 +1,107 @@
+
+#include "at91_clk.h"
+
+#include <dt-bindings/clock/at91.h>
```
move to `#include <>` list, in alphabetical order.

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_pmc.c:
```diff
@@ -0,0 +1,107 @@
+				  num_clks * sizeof(struct pmc_clk);
+	struct pmc_data *pmc_data = NULL;
+
+	pmc_data = malloc(alloc_size);
```
`calloc()`

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_pmc.c:
```diff
@@ -0,0 +1,107 @@
+	for (i = 0; i < nclk; i++) {
+		if (clks[i].clk && clks[i].id == id)
+			return clks[i].clk;
+	}
```
```c
	for (i = 0; i < nclk; i++)
		if (clks[i].id == id)
			return clks[i].clk;

	return NULL;
```
is fine

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_clk.h:
```diff
@@ -0,0 +1,212 @@
+#ifndef AT91_CLK_H
+#define AT91_CLK_H
+
+#include "at91_pmc.h"
```
move local headers after `<>` ones.

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/sama5d2.c:
```diff
@@ -0,0 +1,592 @@
+	.divisors = { 1, 2, 4, 3 },
+};
+
+static uint8_t plla_out[] = { 0 };
```
no need to explicitly zero-init these variables. (here a line 35)

etienne-lms commented on core/arch/arm/plat-sam/drivers/clk/at91_pmc.h:
```diff
@@ -0,0 +1,260 @@
+ * Based on AT91RM9200 datasheet revision E.
+ */
+
+#ifndef AT91_PMC_H
```
`DRIVERS_CLK_AT91_PMC_H`

etienne-lms reviewed (commented):
Also agree with @jenswi-linaro, prefer locate source file in _core/drivers/clk/_.

clementleger commented on core/arch/arm/plat-sam/drivers/clk/at91_main.c:
```diff
@@ -0,0 +1,322 @@
+{
+	uint32_t status = io_read32(osc->base + AT91_PMC_SR);
+
+	return !!(status & AT91_PMC_MOSCRCS);
```
Actually, I think this could be a problem to remove the `!!`.
bool type can be on 8 bit only. when masking  `status & AT91_PMC_MOSCRCS (BIT (17))`, GCC can infer that the lowest bits will always be 0 and thus optimize to always return false.

jforissier replied:
@clementleger no. C99 standard section 6.3.1.2  Boolean type: "When any scalar value is converted to _Bool, the result is 0 if the value compares equal to 0; otherwise, the result is 1".

clementleger replied:
@jforissier: Hum ok, my experience was probably with a typedefed bool that was not mapping the _Bool keyword. Anyway, thanks for the information about C99 _Bool. Regarding this code, I still think it was made for that purpose, since other check are using BIT(0) without !! (`pmc_main_osc_ready`). I will remove them since not needed

clementleger commented on core/arch/arm/plat-sam/drivers/clk/at91_main.c:
```diff
@@ -0,0 +1,322 @@
+	}
+
+	while (!pmc_main_rc_osc_ready(osc))
+		;
```
So, this reveals a "little" problem on the sama5d2 processor. There is no generic timer extension on the processor so all the `read_counter()` and such won't work but will generate an exception. Supporting timeouts would require to abstract the counter used by OP-TEE to use another one provided by a peripheral (TCB) (which relies on clock enabling...).

clementleger commented on core/arch/arm/plat-sam/drivers/clk/at91_main.c:
```diff
@@ -0,0 +1,322 @@
+	}
+
+	while (!pmc_main_osc_ready(pmc))
+		;
```
Ditto

clementleger commented on core/arch/arm/plat-sam/drivers/clk/at91_main.c:
```diff
@@ -0,0 +1,322 @@
+			return TEE_SUCCESS;
+	} while (1);
+
+	panic();
```
I guess not

clementleger commented on core/arch/arm/plat-sam/drivers/clk/at91_main.c:
```diff
@@ -0,0 +1,322 @@
+		mcfr = io_read32(base + AT91_CKGR_MCFR);
+		if (mcfr & AT91_PMC_MAINRDY)
+			return TEE_SUCCESS;
+	} while (1);
```
Ditto (generic timer missing)

etienne-lms commented on core/drivers/sam/at91_audio_pll.c:
```diff
@@ -0,0 +1,420 @@
+			AT91_PMC_AUDIO_PLL_RESETN, 0);
+	io_clrsetbits32(frac->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_RESETN,
+			AT91_PMC_AUDIO_PLL_RESETN);
```
```c
	io_clrbits32(frac->base + AT91_PMC_AUDIO_PLL0,
		     AT91_PMC_AUDIO_PLL_RESETN);
	io_setbits32(frac->base + AT91_PMC_AUDIO_PLL0,
		     AT91_PMC_AUDIO_PLL_RESETN);
```
would be easier to read out.

etienne-lms commented on core/drivers/sam/at91_audio_pll.c:
```diff
@@ -0,0 +1,420 @@
+{
+	struct clk_audio_frac *frac = clk->priv;
+
+	io_clrsetbits32(frac->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_PLLEN, 0);
+	/* do it in 2 separated writes */
+	io_clrsetbits32(frac->base + AT91_PMC_AUDIO_PLL0,
+			AT91_PMC_AUDIO_PLL_RESETN, 0);
```
```c
	io_clrbits32(frac->base + AT91_PMC_AUDIO_PLL0,
		     AT91_PMC_AUDIO_PLL_PLLEN);
	/* Requires 2 separated writes */
	io_clrbits32(frac->base + AT91_PMC_AUDIO_PLL0,
		     AT91_PMC_AUDIO_PLL_RESETN);

etienne-lms commented on core/drivers/sam/at91_audio_pll.c:
```diff
@@ -0,0 +1,420 @@
+	unsigned long long rem = 0;
+
+	if (!rate)
+		return -1;
```
`TEE_ERROR_GENERIC` or another TEE_Result compliant code.

clementleger replied:
used TEE_ERROR_BAD_PARAMETERS since rate is a param

etienne-lms commented on core/drivers/sam/at91_audio_pll.c:
```diff
@@ -0,0 +1,420 @@
+	struct clk_audio_pad *apad_ck = clk->priv;
+	uint8_t tmp_div = 1;
+
+	DMSG("A PLL/PAD: %s, rate = %lu (parent_rate = %lu)\n", __func__, rate,
```
drop argument `__func__` adn trailing `\n`, already handled by `DMSG()`.

clementleger replied:
Done for all DMSG

etienne-lms commented on core/drivers/sam/at91_pmc.h:
```diff
@@ -0,0 +1,260 @@
+ */
+
+#ifndef DRIVERS_CLK_AT91_PMC_H
+#define DRIVERS_CLK_AT91_PMC_H
```
This is a local function, `AT91_PMC_H` would be enough.
Otherwise, prefer stick to header file path: `DRIVERS_SAM_AT91_PM_H`

clementleger replied:
Acked

etienne-lms reviewed (commented):
Could the clock drivers source file be located in _core/drivers/clk/_ directory?
Since there are many source file, maybe even in _core/drivers/clk/sam/_.

clementleger commented:
> Could the clock drivers source file be located in _core/drivers/clk/_ directory? Since there are many source file, maybe even in _core/drivers/clk/sam/_.

Ok, please tell me which path you want precisely and I'll do it.

clementleger commented:
New version: 
- Fix header naming
- Fix some ioclrsetbit with 0 value
- Fix __func__ and \n usage in DMSG

etienne-lms commented:
I think _core/drivers/clk/sam/_ would be a nice place, but I prefer to rely on maintainers point of view.
@jenswi-linaro, @jforissier, @ruchi393? Would you prefer platform drivers in _core/driver/\<platform-name>/_ (as in TF-A) or in drivers specific directories (here _core/drivers/clk/_ if few source files, or _core/drivers/clk/\<platform-name>/_ when many source files as this P-R introduces)?

jforissier commented:
> I think _core/drivers/clk/sam/_ would be a nice place

+1, not a strong opinion but I prefer _clk/sam_ to _sam/clk_ and that is how the Linux kernel does it too.

clementleger commented:
New version:
- Fixed some trivial checkpatch warnings. A lot of them are potentially better not be fixed (some long lines)
- Moved everything in `core/drivers/clk/sam`

jenswi-linaro commented:
> > I think _core/drivers/clk/sam/_ would be a nice place
> 
> +1, not a strong opinion but I prefer _clk/sam_ to _sam/clk_ and that is how the Linux kernel does it too.

+1

clementleger commented:
Ping ? There are still checkpatch warnings but they are almost all "CHECKS" or long lines that can hardly be removed or it'll be less readable/maintainable.

jforissier commented:
@clementleger the CHECK messages about macro reuse are probably OK to ignore although they could easily be changed with temporary variables with the `({ ... })` syntax. @jenswi-linaro what do you think? We have a bunch of macros that triggered this CHECK already.

As for the line too long warnings, they're all easy to fix aren't they? (just indent one or two tabs less).

clementleger commented:
> As for the line too long warnings, they're all easy to fix aren't they? (just indent one or two tabs less).

I'll try that but last time I did (maybe not exactly the same problem), `checkpatch` then displayed that I wasn't aligned anymore on parenthesis.

jenswi-linaro commented:
> @clementleger the CHECK messages about macro reuse are probably OK to ignore although they could easily be changed with temporary variables with the `({ ... })` syntax. @jenswi-linaro what do you think? We have a bunch of macros that triggered this CHECK already.

It's not that hard to fix the macro CHECKs as you describe it. So let's just fix those too, unless there's a good reason to why the macros should be kept as is.

clementleger commented:
> > @clementleger the CHECK messages about macro reuse are probably OK to ignore although they could easily be changed with temporary variables with the `({ ... })` syntax. @jenswi-linaro what do you think? We have a bunch of macros that triggered this CHECK already.
> 
> It's not that hard to fix the macro CHECKs as you describe it. So let's just fix those too, unless there's a good reason to why the macros should be kept as is.

Ok I'll try to fix most of them. Regarding this one, I have absolutely no idea why checkpatch triggers that:
```
ERROR: Macros with complex values should be enclosed in parentheses
#267: FILE: core/drivers/clk/sam/at91_pmc.h:244:
+#define          AT91_PMC_AUDIO_PLL_QDPAD(n) \
+                               ((n) << AT91_PMC_AUDIO_PLL_QDPAD_OFFSET)

```
And `AT91_PMC_AUDIO_PLL_QDPAD_OFFSET` is defined this way:

`#define	  AT91_PMC_AUDIO_PLL_QDPAD_OFFSET	24`

etienne-lms commented on core/drivers/clk/sam/sama5d2_clk.c:
```diff
@@ -0,0 +1,592 @@
+	struct clk_range r;
+} sama5d2_peri32ck[] = {
+	{
+		.clk = {.n = "macb0_clk", .id = 5 },
```
add a space char after `{`, here and in below lines

etienne-lms commented on core/drivers/clk/sam/sama5d2_clk.c:
```diff
@@ -0,0 +1,592 @@
+				ARRAY_SIZE(sama5d2_gck),
+				ARRAY_SIZE(sama5d2_progck));
+	pmc->base = base;
+	if (!pmc)
```
```patch
-	pmc->base = base;
	if (!pmc)
 		panic();
+	pmc->base = base;
```
or
```patch
+	assert(pmc);
 	pmc->base = base;
-	if (!pmc)
-		panic();
```

clementleger replied:
Ouch, nasty one :/

etienne-lms commented on core/drivers/clk/sam/sama5d2_clk.c:
```diff
@@ -0,0 +1,592 @@
+				    &sama5d3_pll_layout, &plla_characteristics);
+	if (!clk)
+		panic();
+	plladivck = at91_clk_register_plldiv(pmc, "plladivck", clk);
```
add an empty line above

etienne-lms reviewed (commented):
Some comments. Note all calls to `malloc()` should be relaced with `calloc()`.

jenswi-linaro commented:
> Ok I'll try to fix most of them. Regarding this one, I have absolutely no idea why checkpatch triggers that:
> 
> ```
> ERROR: Macros with complex values should be enclosed in parentheses
> #267: FILE: core/drivers/clk/sam/at91_pmc.h:244:
> +#define          AT91_PMC_AUDIO_PLL_QDPAD(n) \
> +                               ((n) << AT91_PMC_AUDIO_PLL_QDPAD_OFFSET)
> ```
> 
> And `AT91_PMC_AUDIO_PLL_QDPAD_OFFSET` is defined this way:
> 
> `#define AT91_PMC_AUDIO_PLL_QDPAD_OFFSET 24`

It looks a bit dangerous, what if the supplied `n` is signed and too large?
I suggest using the `SHIFT_U32()` macro.

> 
> Should I also fix the false positive such as the following ?
> 
> ```
> 
> WARNING: 'mor' may be misspelled - perhaps 'more'?
> #166: FILE: core/drivers/clk/sam/at91_main.c:142:
> +               io_write32(pmc->base + AT91_CKGR_MOR, mor);
> ```

False positives can be ignored.

clementleger commented on core/drivers/clk/sam/at91_audio_pll.c:
```diff
@@ -0,0 +1,418 @@
+	fout = clk_audio_pll_fout(parent_rate, frac->nd, frac->fracr);
+
+	DMSG("A PLL: fout = %lu (nd = %u, fracr = %lu)", fout, frac->nd,
+	     (unsigned long)frac->fracr);
```
Removed debug entirely.

clementleger commented on core/drivers/clk/sam/at91_audio_pll.c:
```diff
@@ -0,0 +1,418 @@
+
+	apmc_rate = parent_rate / (apmc_ck->qdpmc + 1);
+
+	DMSG("A PLL/PMC: apmc_rate = %lu (qdpmc = %u)", apmc_rate,
```
Idem, removed debug.

clementleger commented on core/drivers/clk/sam/at91_main.c:
```diff
@@ -0,0 +1,322 @@
+		mcfr = io_read32(base + AT91_CKGR_MCFR);
+		if (mcfr & AT91_PMC_MAINRDY)
+			return TEE_SUCCESS;
+	} while (1);
```
Indeed

clementleger commented on core/drivers/clk/sam/at91_main.c:
```diff
@@ -0,0 +1,322 @@
+			return TEE_SUCCESS;
+	} while (1);
+
+	return TEE_SUCCESS;
```
Need with your previous change

clementleger commented:
> > And `AT91_PMC_AUDIO_PLL_QDPAD_OFFSET` is defined this way:
> > `#define AT91_PMC_AUDIO_PLL_QDPAD_OFFSET 24`
> 
> It looks a bit dangerous, what if the supplied `n` is signed and too large? I suggest using the `SHIFT_U32()` macro.

Ok, thanks for the macro !

clementleger commented:
New version:
- Removed debug
- Used SHIFT_U32
- Fixed a few macros which used parameters twice

jenswi-linaro commented:
Let us know when there's only what you consider false positives left in the checkpatch complaints.

clementleger commented:
@jenswi-linaro: Currently, there are only 80 lines warning that I did not found a sensible way to fix. If I move them, there is a warning about the parenthesis alignment.

Another one is this one:
```
ERROR: Macros with complex values should be enclosed in parentheses
#2457: FILE: core/drivers/clk/sam/at91_pmc.h:244:
+#define	  AT91_PMC_AUDIO_PLL_QDPAD(n) \
+				SHIFT_U32(n, AT91_PMC_AUDIO_PLL_QDPAD_OFFSET)
```
For some unknown reason, checkpatch think this macro uses a complex value.

I will check the build failure error that I can't reproduce on my side.

clementleger commented:
> The problem seems to be that `characteristics` is a quite long word, it must be quite painful to use as a name of a variable or even as part of a name of a struct for that matter. It seems this line length limit has the nice side effect that it catches some unreasonably long names.

Meh, let's rename all of that.

clementleger commented:
New version:
- Rename peripheral -> periph when needed
- Rename characteristics -> charac

jenswi-linaro commented on core/drivers/clk/sam/at91_master.c:
```diff
@@ -0,0 +1,185 @@
+	struct clk_master *master = clk->priv;
+	const struct clk_master_layout *layout = master->layout;
+	const struct clk_master_charac *charac =
+						master->charac;
```
Please fold up here and at the other places in this PR.

etienne-lms commented on core/drivers/clk/sam/at91_programmable.c:
```diff
@@ -0,0 +1,163 @@
+({ \
+	typeof(layout) __layout = layout; \
+	(((pckr) >> (__layout)->pres_shift) & (__layout)->pres_mask); \
+})
```
Indentation (+ should be an empty line between variable definitions and instructions)
```c
#define PROG_PRES(layout, pckr)	\
	({ \
		typeof(layout) __layout = layout; \
		\
		(((pckr) >> (__layout)->pres_shift) & (__layout)->pres_mask); \
	})

etienne-lms commented on core/drivers/clk/sam/at91_programmable.c:
```diff
@@ -0,0 +1,163 @@
+	struct clk *clk = NULL;
+
+	if (id > PROG_ID_MAX)
+		return NULL;
```
should not happen, right?  Assert or panic?

clementleger replied:
Should never happen, adding an `assert()`.

etienne-lms commented on core/drivers/clk/sam/at91_sckc.c:
```diff
@@ -0,0 +1,53 @@
+#include <libfdt.h>
+#include <malloc.h>
+#include <stdint.h>
+#include <util.h>
```
I think some header files could be removed here

etienne-lms commented on core/drivers/clk/sam/at91_utmi.c:
```diff
@@ -0,0 +1,138 @@
+		io_clrsetbits32(utmi->sfr_base + AT91_SFR_UTMICKTRIM,
+				AT91_UTMICKTRIM_FREQ, utmi_ref_clk_freq);
+	} else if (utmi_ref_clk_freq) {
+		EMSG("UTMICK: sfr node required\n");
```
remove `\n`

clementleger replied:
Fixed here and in `at91_main` also.

etienne-lms commented on core/drivers/clk/sam/at91_utmi.c:
```diff
@@ -0,0 +1,138 @@
+{
+	struct clk_utmi *utmi = clk->priv;
+
+	io_clrsetbits32(utmi->pmc_base + AT91_CKGR_UCKR, AT91_PMC_UPLLEN, 0);
```
`io_clrbits32()`

etienne-lms commented on core/drivers/clk/sam/at91_pmc.c:
```diff
@@ -0,0 +1,111 @@
+struct clk *clk_dt_pmc_get(struct dt_driver_phandle_args *clkspec, void *data,
+			   TEE_Result *res)
+{
+	unsigned int type = clkspec->args[0];
```
should check `clkspec->args_count == 2`

etienne-lms commented on core/drivers/clk/sam/at91_clk.h:
```diff
@@ -0,0 +1,222 @@
+({ \
+	typeof(_mask) __mask = _mask; \
+	(((_val) << (ffs(__mask) - 1)) & (__mask)); \
+})
```
indentation

clementleger replied:
Fixed + empty line after decl.

etienne-lms commented on core/drivers/clk/sam/sama5d2_clk.c:
```diff
@@ -0,0 +1,594 @@
+#include <assert.h>
+#include <initcall.h>
+#include <kernel/boot.h>
+#include <libfdt.h>
```
move below line 10

clementleger replied:
Remove initcall.h entirely since not needed anymore.

etienne-lms commented on core/drivers/clk/sub.mk:
```diff
@@ -1,3 +1,5 @@
 srcs-$(CFG_DRIVERS_CLK_DT) += clk_dt.c
 srcs-$(CFG_DRIVERS_CLK_FIXED) += fixed_clk.c
+
+subdirs-y += sam
```
i think it should better be here:
`subdirs-$(CFG_DRIVERS_SAM_CLK) += sam`

And in _clk/sam/sub.mk_ use a specific `CFG_SAMA5D2_CLK` config switch for _sama5d2_clk.c_ (or unconditionally embedded it since sama5d2 is the only existing flavor/variant of plat-sam)

clementleger replied:
Ok

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "ext: include: add DIV_ROUND_UP function"

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "dts: at91: removed useless definition"
* "drivers: sam: add at91_audio_pll clock driver"
* "drivers: sam: add at91_h32mx clock driver"
* "drivers: sam: add at91_i2s_mux clock driver"
* "drivers: sam: add at91_peripheral clock driver" with minor comments
* "drivers: sam: add at91_pll clock driver" with minor comments
* "drivers: sam: add at91_plldiv clock driver"
* "drivers: sam: add at91_programmable clock driver" with minor comments
* "drivers: sam: add at91_usb clock driver"
* "drivers: sam: add at91_utmi clock driver" with minor comments
* "drivers: sam: add at91_system clock driver"
* "drivers: sam: add pmc clock registering" with minor comment
* "drivers: sam: add at91 clock interface" with minor comment
* "drivers: sam: add sama5d2 clock description" with a minor comment

clementleger commented on core/drivers/clk/sam/at91_pmc.h:
```diff
@@ -0,0 +1,260 @@
+#define	AT91_CKGR_PLLBR				0x2c
+#define	  AT91_PMC_DIV				(0xff  <<  0)
+#define	  AT91_PMC_PLLCOUNT			(0x3f  <<  8)
+#define	  AT91_PMC_OUT				(3     << 14)
```
Chekcpatch did not trigger on this but I fixed them all after your comment.

clementleger commented on core/drivers/clk/sam/at91_pmc.h:
```diff
@@ -0,0 +1,260 @@
+#define AT91_PMC_CPU_CKR			0x28
+
+#define	AT91_PMC_MCKR				0x30
+#define	  AT91_PMC_CSS				(3 <<  0)
```
Fixed also everywhere in this file.

clementleger commented on core/drivers/clk/sam/at91_main.c:
```diff
@@ -0,0 +1,316 @@
+
+	/* Enable the oscillator if not */
+	if (!(mor & AT91_PMC_MOSCRCEN)) {
+		mor &= ~(MOR_KEY_MASK | AT91_PMC_MOSCRCEN);
```
Indeed.

clementleger commented on core/drivers/clk/sam/at91_master.c:
```diff
@@ -0,0 +1,185 @@
+
+	pres = (val >> master->layout->pres_shift) & MASTER_PRES_MASK;
+	if (pres == 3 && charac->have_div3_pres)
+		pres = 3;
```
Fixed with proposed line

clementleger commented on core/drivers/clk/sam/at91_peripheral.c:
```diff
@@ -0,0 +1,189 @@
+#include "at91_clk.h"
+
+#define PERIPHERAL_ID_MIN	2
+#define PERIPHERAL_ID_MAX	31
```
Agreed

clementleger commented on core/drivers/clk/sam/at91_peripheral.c:
```diff
@@ -0,0 +1,189 @@
+		if (!parent_rate)
+			return;
+
+		for (; shift < PERIPHERAL_MAX_SHIFT; shift++) {
```
Added `shift = 0`

clementleger commented on core/drivers/clk/sam/at91_generated.c:
```diff
@@ -0,0 +1,180 @@
+	io_write32(gck->base + gck->layout->offset,
+		   (gck->id & gck->layout->pid_mask));
+	io_clrsetbits32(gck->base + gck->layout->offset,
+			gck->layout->cmd | AT91_PMC_PCR_GCKEN,
```
Actually, `gck->layout->cmd` is a single bit value on, this platform. The goal of this (I'm guessing) was to simply clear the bit `AT91_PMC_PCR_GCKEN`. So i think the following code is actually sufficient:
```
	io_clrsetbits32(gck->base + gck->layout->offset,
			AT91_PMC_PCR_GCKEN, gck->layout->cmd);
```

clementleger commented:
New version:
- Added Acked-by and Reviewed-By from Etienne

jforissier commented on core/drivers/clk/sam/sub.mk:
```diff
@@ -5,4 +5,4 @@ at91-common += at91_utmi.c at91_master.c at91_h32mx.c at91_usb.c
 at91-common += at91_generated.c at91_i2s_mux.c at91_audio_pll.c
 
-srcs-$(CFG_SAMA5D2_CLK) += $(at91-common) sama5d2_clk.c
+srcs-$(CFG_DRIVERS_SAMA5D2_CLK) += $(at91-common) sama5d2_clk.c
```
Why not use this CFG name in the parent commit? ("drivers: add sub.mk for sam")

clementleger replied:
That's an error, I made a fixup on the wrong commit. This one should be in the previous commit. I used `CFG_DRIVERS_SAMA5D2_CLK` since I think it more coherent with the other CFG names, tell me if you want to keep `CFG_SAMA5D2_CLK` for some reason.
